src/core/tsi/alts/crypt/aes_gcm.cc

```diff
@@ -35,26 +39,113 @@ constexpr size_t kKdfCounterLen = 6;
 constexpr size_t kKdfCounterOffset = 2;
 constexpr size_t kRekeyAeadKeyLen = kAes128GcmKeyLength;
 
+namespace grpc_core {
+
+GsecKeyFactory::GsecKeyFactory(absl::Span<const uint8_t> key, bool is_rekey)
+    : key_(std::vector<uint8_t>(key.begin(), key.end())), is_rekey_(is_rekey) {}
+
+std::unique_ptr<GsecKeyInterface> GsecKeyFactory::Create() const {
+  return std::make_unique<GsecKey>(absl::MakeConstSpan(key_), is_rekey_);
+}
+
+GsecKey::GsecKey(absl::Span<const uint8_t> key, bool is_rekey)
+    : is_rekey_(is_rekey) {
+  if (is_rekey_) {
+    aead_key_ = std::vector<uint8_t>(kRekeyAeadKeyLen);
+    kdf_buffer_ = std::vector<uint8_t>(EVP_MAX_MD_SIZE);
+    nonce_mask_ = std::vector<uint8_t>(kAesGcmNonceLength);
+    memcpy(nonce_mask_.data(), key.data() + kKdfKeyLen, kAesGcmNonceLength);
+    kdf_counter_ = std::vector<uint8_t>(kKdfCounterLen, 0);
+  }
+  key_ = std::vector<uint8_t>(is_rekey_ ? kKdfKeyLen : key.size());
+  memcpy(key_.data(), key.data(), key_.size());
```

Comment:
Is this line correct? Should it be `is_rekey_ ? kKdfKeyLen : key.size()` instead?

If yes, can we add a unit test that would have caught this?

Reply:
Ah of course! Sorry I missed this. :)